From f91b0cf38f640a031de08e27081599da13be92fe Mon Sep 17 00:00:00 2001
From: Andrew Chow <achow101-github@achow101.com>
Date: Tue, 27 Nov 2018 17:32:44 -0500
Subject: [PATCH 1/3] Use linux unix socket address format

---
 unix/frozen-modules/pyb.py | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/unix/frozen-modules/pyb.py b/unix/frozen-modules/pyb.py
index 5341be7..b9d32d0 100644
--- a/unix/frozen-modules/pyb.py
+++ b/unix/frozen-modules/pyb.py
@@ -26,10 +26,10 @@ class USB_HID:
         fn = b'/tmp/ckcc-simulator.sock'
         self.pipe = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
         # If on linux, try commenting the following line
-        addr = bytes([len(fn)+2, socket.AF_UNIX] + list(fn))
+        # addr = bytes([len(fn)+2, socket.AF_UNIX] + list(fn))
         # If on linux, try uncommenting the following two lines
-        #import struct
-        #addr = struct.pack('H108s', socket.AF_UNIX, fn)
+        import struct
+        addr = struct.pack('H108s', socket.AF_UNIX, fn)
         while 1:
             try:
                 self.pipe.bind(addr)
-- 
2.30.0


From 32d4235a79e8cde7ee1c9901e0c834c310a29725 Mon Sep 17 00:00:00 2001
From: Andrew Chow <achow101-github@achow101.com>
Date: Tue, 17 Dec 2019 17:56:05 -0500
Subject: [PATCH 2/3] Change default simulator multisig

---
 unix/frozen-modules/sim_settings.py | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/unix/frozen-modules/sim_settings.py b/unix/frozen-modules/sim_settings.py
index 650f255..cf5bbd4 100644
--- a/unix/frozen-modules/sim_settings.py
+++ b/unix/frozen-modules/sim_settings.py
@@ -68,7 +68,11 @@ if '--ms' in sys.argv:
         sim_defaults['multisig'] = [["CC-2-of-4", [2, 4], [[1130956047, "tpubDF2rnouQaaYrUEy2JM1YD3RFzew4onawGM4X2Re67gguTf5CbHonBRiFGe3Xjz7DK88dxBFGf2i7K1hef3PM4cFKyUjcbJXddaY9F5tJBoP"], [3503269483, "tpubDFcrvj5n7gyatVbr8dHCUfHT4CGvL8hREBjtxc4ge7HZgqNuPhFimPRtVg6fRRwfXiQthV9EBjNbwbpgV2VoQeL1ZNXoAWXxP2L9vMtRjax"], [2389277556, "tpubDExj5FnaUnPAjjgzELoSiNRkuXJG8Cm1pbdiA4Hc5vkAZHphibeVcUp6mqH5LuNVKbtLVZxVSzyja5X26Cfmx6pzRH6gXBUJAH7MiqwNyuM"], [3190206587, "tpubDFiuHYSJhNbHaGtB5skiuDLg12tRboh2uVZ6KGXxr8WVr28pLcS7F3gv8SsHFa2tm1jtx3VAuw56YfgRkdo6DXyfp51oygTKY3nJFT5jBMt"]], {"pp": "48'/1'/0'/1'", "ch": "XTN", "ft": 26}]]
     else:
         # P2SH: 2of4 using BIP39 passwords: "Me", "Myself", "and I", and (empty string) on simulator
-        sim_defaults['multisig'] = [['MeMyself', [2, 4], [[3503269483, 'tpubD9429UXFGCTKJ9NdiNK4rC5ygqSUkginycYHccqSg5gkmyQ7PZRHNjk99M6a6Y3NY8ctEUUJvCu6iCCui8Ju3xrHRu3Ez1CKB4ZFoRZDdP9'], [2389277556, 'tpubD97nVL37v5tWyMf9ofh5rznwhh1593WMRg6FT4o6MRJkKWANtwAMHYLrcJFsFmPfYbY1TE1LLQ4KBb84LBPt1ubvFwoosvMkcWJtMwvXgSc'], [3190206587, 'tpubD9ArfXowvGHnuECKdGXVKDMfZVGdephVWg8fWGWStH3VKHzT4ph3A4ZcgXWqFu1F5xGTfxncmrnf3sLC86dup2a8Kx7z3xQ3AgeNTQeFxPa'], [1130956047, 'tpubD8NXmKsmWp3a3DXhbihAYbYLGaRNVdTnr6JoSxxfXYQcmwVtW2hv8QoDwng6JtEonmJoL3cNEwfd2cLXMpGezwZ2vL2dQ7259bueNKj9C8n']], {'ch': 'XTN', 'pp': "45'"}]]
+        sim_defaults['multisig'] = [
+            ['mstest', [2, 3], [[1130956047, 0, 'tpubDF2rnouQaaYrR9x68P5Jm8WjhCE4atyGiPviFA9ve5iMnYbkTjof2HjzejcQcD7getPusDLPsWJLN2UttzK3pyVgBkRs52MiRZM7ZJ8TrEq'], [1130956047, 1, 'tpubDETRnZNJAqXiVeiL8UMDzCTBAoh3JvZkgXLdb1K2xzpJLepuJ6ka8jnVyRSkVh8Nbbo8u8dobZCsNENmRKipLzHNsS5mccjKSpXgSgavTQe'], [1130956047, 2, 'tpubDF3hdPQ5oDhtYjjaC596pboPii7UZmjqZcBPBRAbb6Bgn9hKoFxb8zWsBfdiCnTq3htUs2Yi2reeG3kMqHzZGZykJQAB5aKrJ8UfiXjmaLD']], {'ft': 8, 'ch': 'XTN', "d": ["48'/1'/0'/0'", "48'/1'/1'/0'", "48'/1'/2'/0'"]}],
+            ['mstest1', [2, 3], [[1130956047, 0, 'tpubDF2rnouQaaYrUEy2JM1YD3RFzew4onawGM4X2Re67gguTf5CbHonBRiFGe3Xjz7DK88dxBFGf2i7K1hef3PM4cFKyUjcbJXddaY9F5tJBoP'], [1130956047, 1, 'tpubDETRnZNJAqXiVeiL8UMDzCTBAoh3JvZkgXLdb1K2xzpJLepuJ6ka8jnVyRSkVh8Nbbo8u8dobZCsNENmRKipLzHNsS5mccjKSpXgSgavTQe'], [1130956047, 2, 'tpubDF3hdPQ5oDhtYjjaC596pboPii7UZmjqZcBPBRAbb6Bgn9hKoFxb8zWsBfdiCnTq3htUs2Yi2reeG3kMqHzZGZykJQAB5aKrJ8UfiXjmaLD']], {'ft': 14, 'ch': 'XTN', "d": ["48'/1'/0'/0'", "48'/1'/1'/0'", "48'/1'/2'/0'"]}],
+            ['mstest2', [2, 3], [[1130956047, 0, 'tpubDF2rnouQaaYrXF4noGTv6rQYmx87cQ4GrUdhpvXkhtChwQPbdGTi8GA88NUaSrwZBwNsTkC9bFkkC8vDyGBVVAQTZ2AS6gs68RQXtXcCvkP'], [1130956047, 1, 'tpubDETRnZNJAqXiVeiL8UMDzCTBAoh3JvZkgXLdb1K2xzpJLepuJ6ka8jnVyRSkVh8Nbbo8u8dobZCsNENmRKipLzHNsS5mccjKSpXgSgavTQe'], [1130956047, 2, 'tpubDF3hdPQ5oDhtYjjaC596pboPii7UZmjqZcBPBRAbb6Bgn9hKoFxb8zWsBfdiCnTq3htUs2Yi2reeG3kMqHzZGZykJQAB5aKrJ8UfiXjmaLD']], {'ft': 26, 'ch': 'XTN', "d": ["48'/1'/0'/0'", "48'/1'/1'/0'", "48'/1'/2'/0'"]}],
+            ]
     sim_defaults['fee_limit'] = -1
 
 if '--xfp' in sys.argv:
-- 
2.30.0


From 8cf7883d84ca5e30638b4aff0cd12c7e9a01e643 Mon Sep 17 00:00:00 2001
From: Andrew Chow <achow101-github@achow101.com>
Date: Wed, 27 Jan 2021 21:50:22 -0500
Subject: [PATCH 3/3] Allow multisigs to share master fingerprint

---
 shared/multisig.py | 38 ++++++++++++++++++++++++--------------
 1 file changed, 24 insertions(+), 14 deletions(-)

diff --git a/shared/multisig.py b/shared/multisig.py
index b7e4f44..e3b5807 100644
--- a/shared/multisig.py
+++ b/shared/multisig.py
@@ -141,9 +141,9 @@ class MultisigWallet:
         # calc useful cache value: numeric xfp+subpath, with lookup
         self.xfp_paths = {}
         for xfp, deriv, _ in self.xpubs:
-            self.xfp_paths[xfp] = str_to_keypath(xfp, deriv)
+            self.xfp_paths.setdefault(xfp, list()).append(str_to_keypath(xfp, deriv))
 
-        assert len(self.xfp_paths) == self.N, 'dup XFP'         # not supported
+        assert len(self.xpubs) == self.N, 'Number of pubkeys does not match N'
 
     @classmethod
     def render_addr_fmt(cls, addr_fmt):
@@ -244,7 +244,11 @@ class MultisigWallet:
 
     def get_xfp_paths(self):
         # return list of lists [xfp, *deriv]
-        return list(self.xfp_paths.values())
+        ret = []
+        for paths_list in self.xfp_paths:
+            for xfp_path in paths_list:
+                ret.append(xfp_path)
+        return ret
 
     @classmethod
     def find_match(cls, M, N, xfp_paths, addr_fmt=None):
@@ -282,17 +286,23 @@ class MultisigWallet:
         for x in xfp_paths:
             if x[0] not in self.xfp_paths:
                 return False
-            prefix = self.xfp_paths[x[0]]
-
-            if len(x) < len(prefix):
-                # PSBT specs a path shorter than wallet's xpub
-                #print('path len: %d vs %d' % (len(prefix), len(x)))
-                return False
-
-            comm = len(prefix)
-            if tuple(prefix[:comm]) != tuple(x[:comm]):
-                # xfp => maps to wrong path
-                #print('path mismatch:\n%r\n%r\ncomm=%d' % (prefix[:comm], x[:comm], comm))
+            for prefix in self.xfp_paths[x[0]]:
+                if len(x) < len(prefix):
+                    # PSBT specs a path shorter than wallet's xpub
+                    #print('path len: %d vs %d' % (len(prefix), len(x)))
+                    return False
+
+                comm = len(prefix)
+                if tuple(prefix[:comm]) != tuple(x[:comm]):
+                    # xfp => maps to wrong path
+                    # But maybe there is another path that does match, so keep going
+                    #print('path mismatch:\n%r\n%r\ncomm=%d' % (prefix[:comm], x[:comm], comm))
+                    continue
+                else:
+                    # Found a match, cleanly exit
+                    break
+            else:
+                # No match was found
                 return False
 
         return True
-- 
2.30.0

