From 3d4ff5c0f7ffa434f73990f03b21c20153d810da Mon Sep 17 00:00:00 2001
From: Andrew Chow <achow101-github@achow101.com>
Date: Fri, 15 Nov 2019 18:20:25 -0500
Subject: [PATCH] Send out screen text over unix socket

---
 mcu/bagl.py | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/mcu/bagl.py b/mcu/bagl.py
index e68ab9c..8c8b3f7 100644
--- a/mcu/bagl.py
+++ b/mcu/bagl.py
@@ -1,4 +1,8 @@
 import binascii
+import json
+import os
+import socket
+
 from collections import namedtuple
 from construct import *
 
@@ -61,6 +65,8 @@ SEPROXYHAL_TAG_SCREEN_DISPLAY_RAW_STATUS_START = 0x00
 
 DrawState = namedtuple('DrawState', 'x y width height colors bpp xx yy')
 
+SCREEN_TEXT_SOCKET = '/tmp/ledger-screen.sock'
+
 class Bagl:
     def __init__(self, m, width, height):
         self.m = m
@@ -69,6 +75,8 @@ class Bagl:
 
         self.draw_state = DrawState(0, 0, 0, 0, [], 0, 0, 0)
 
+        self.screen_text_sock = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+
     def refresh(self):
         self.m.update()
 
@@ -566,6 +574,11 @@ class Bagl:
         elif type_ == BAGL_LABEL:
             self._display_bagl_labeline(component, context, halignment, valignment, baseline, char_height, strwidth, type_)
         elif type_ == BAGL_LABELINE:
+            try:
+                data = {"y": component.y, "text": context.decode()}
+                self.screen_text_sock.sendto(json.dumps(data).encode(), SCREEN_TEXT_SOCKET)
+            except:
+                pass
             self._display_bagl_labeline(component, context, halignment, valignment, baseline, char_height, strwidth, type_)
         elif type_ == BAGL_ICON:
             self._display_bagl_icon(component, context)
-- 
2.24.0

